File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public void doubly(Node head){
        Node ptr=head,forw=head.next;
        while(ptr!=null){
            Node new_n = new Node(ptr.val);
            forw=ptr.next;
            ptr.next=new_n;
            new_n.next=forw;
            ptr=forw;
        }
    }
    public Node singly(Node head){
        Node ptr=head,forw=head.next,temp;
        head=head.next;
        
        while(forw!=null){
            
            ptr.next=forw.next;
            if(forw.next!=null)
                forw.next=forw.next.next;
            else
                forw.next=null;
            ptr=ptr.next;
            forw=forw.next;

        }
        return head;
    }
    public Node copyRandomList(Node head) {
        if(head==null)
            return head;
        doubly(head);
        Node ran=null,ptr=head;
        while(ptr!=null){
            if(ptr.random==null){
                ptr.next.random=null;
            }
            else{
                ptr.next.random=ptr.random.next;
            }
            ptr=ptr.next.next;
        }

        return singly(head);
    }
}

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':

        d={}
        ptr=head
        head2=None

        while(ptr!=None):
            temp = Node (ptr.val)
            d[ptr]=temp
            if head2==None:
                head2=temp
            else:
                ptr2=head2
                while ptr2.next!=None:
                    ptr2=ptr2.next
                ptr2.next=temp
            ptr=ptr.next
        ptr=head
        ptr2=head2
        while ptr!=None:
            if(ptr.random==None):
                ptr2.random=None
            else:
                ptr2.random=d[ptr.random]
            ptr=ptr.next
            ptr2=ptr2.next
        return head2

        
        